### ISSUE TITLE
A Server Cookies Store for long cookies

### ISSUE THREAD

**Issue by brianv0** [pusher/oauth2_proxy#138]
One great feature about oauth2_proxy is that it doesn't need local storage, as tokens and additional information is stored in cookies. From a security perspective, this is also ideal, as there's no store to worry about compromising a session store and accessing tokens. 

This can lead to large cookies, especially for OpenID Connect providers that include a lot of claims in a token. It's known that long cookies (> 4kB) can cause some issues (#69) although that can mitigated by splitting cookies, though this can introduce additional complexities (#29), although those can be mitigated as well.

However, the data sent over the wire is still large and doesn't compress well since it's encrypted.

## Expected Behavior

Over the wire cookie size should be kept to a minimum if possible.

## Current Behavior

Cookies are large over the wire for some OIDC providers.

## Possible Solution

Implementing a Server Cookies Store would alleviate that. Instead of sending the user a cookie whose value is the encoded, encrypted session, oauth2_proxy would send the user a ticket for that same cookie. That ticket would include information about how to retrieve the cookie, a _handle_. In addition to that, to preserve the current security aspect of oauth2_proxy, that ticket would also include information necessary to decrypt the cookie value.

When issuing a cookie, the handle would effectively be a random number to prevent collisions. The information necessary to decrypt the cookie can be a key or a random 16-byte initialization vector like is used to currently encrypt the other fields.

When refreshing a session, you would use that same ticket and just overwrite the value.

Redis is also an ideal storage backend for this, as you can also use session expiration as key expiration.

## Context

Nginx configuration has been a bit problematic with large tokens. This doesn't really make this go away if you are also setting the Authorization header again, for example, but it may help in some cases.

In addition to that, with some changes #65, it is possible to connect simpler HTTP clients up with a token and HTTP Basic, such as the WebDAV client included in Mac OS X. However, the cookies are too large for some of these old clients which expect username or password combinations to be less than 256 bytes.

Finally, for applications that make many small requests, 4kB of additional information that is currently sent can be reduces to less than 100, which may make for a big improvement for chatty applications with large amounts of users.

I have an implementation that can be layered on top of the current master (46a0d7c) as well as changes from #65 should the be merged.

The majority of that work can be seen in this commit: c3e30ca, additional commits contain unit tests and support for tickets in the Authorization header for #65 as well.

**Comment by JoelSpeed** [comment]
@brianv0 I've been considering something like this for a while! Thanks for the thorough description 😄 

Could you open a PR for the changes or does #65 need merging first? 

Does this implementation mean that the client-side cookie is only updated when the server side stored cookie has gone away? IE provided refreshing is available, you should only get a `set-cookie` header very very rarely when you have been logged out for some reason or you haven't logged in for a while? 

I haven't done a thorough review of the code but a quick scan looks initially good. One thought I have had is that maybe the current cookie implementation could also become a `CookieStore` of some description so that you would have either the `ClientCookieStore` or `RedisCookieStore` enabled and there would be a nice interface in front of the pair of them, moving all of the cookie logic into a separate package and then the main OAuth2 code could just call `cookieStore.SaveCookie()` or `cookieStore.SaveSession()` or something like that, which would return an `*http.Cookie` that should be sent to the client? Does this make sense? 

Perhaps this needs some deeper thought 🤔

**Comment by brianv0** [comment]
Could you open a PR for the changes or does #65<https://github.com/pusher/oauth2_proxy/pull/65> need merging first?

I can open a PR, the commits are basically split by feature.


Does this implementation mean that the client-side cookie is only updated when the server side stored cookie has gone away? IE provided refreshing is available, you should only get a set-cookie header very very rarely when you have been logged out for some reason or you haven't logged in for a while?

Yes, Set-Cookie can happen rarely in this scenario. If the Server Side cookie has disappeared, it's effectively a new login (the same as an expired cookie now), so you get a new cookie. I originally implemented a FileSystemCookieStore before the Redis one but I removed that code since I didn't have a way of cleaning up expired cookies from abandoned browser sessions (`SETEX` in redis takes care of this nicely).

I haven't done a thorough review of the code but a quick scan looks initially good. One thought I have had is that maybe the current cookie implementation could also become a CookieStore of some description so that you would have either the ClientCookieStore or RedisCookieStore enabled and there would be a nice interface in front of the pair of them, moving all of the cookie logic into a separate package and then the main OAuth2 code could just call cookieStore.SaveCookie() or cookieStore.SaveSession() or something like that, which would return an *http.Cookie that should be sent to the client? Does this make sense?

Yeah, I had toyed around with that same implementation also at first, though I think you actually need to return an array of cookies in order to do this right and handle the split cookies I think, and/or your interface needs to accept request objects for Store/Load/Clear instead of simpler objects. You also need to move around code and configuration from oauthproxy.go to the cookie store. Ultimately I decided to take a very conservative approach on modifying the existing code while still adding this feature.

I can push around some code today to see if I can come up with an alternative implementation and what that could look like, though I think it will involve modifying the existing source code a bit.

**Comment by JoelSpeed** [comment]
> Yes, Set-Cookie can happen rarely in this scenario. If the Server Side cookie has disappeared, it's effectively a new login (the same as an expired cookie now), so you get a new cookie. I originally implemented a FileSystemCookieStore before the Redis one but I removed that code since I didn't have a way of cleaning up expired cookies from abandoned browser sessions (`SETEX` in redis takes care of this nicely).

This sounds like a great improvement! I do also like the idea of `FileSystemCookieStore` for those who don't want to set up a redis but alas, you are correct, reaping sessions would be difficult. I'm sure it could be done but perhaps not easily.

> Yeah, I had toyed around with that same implementation also at first, though I think you actually need to return an array of cookies in order to do this right and handle the split cookies I think, and/or your interface needs to accept request objects for Store/Load/Clear instead of simpler objects.

Taking the request object in the interface might be the simplest way to do this 🤔 

> Ultimately I decided to take a very conservative approach on modifying the existing code while still adding this feature.

I appreciate that :+1:

> I can push around some code today to see if I can come up with an alternative implementation and what that could look like, though I think it will involve modifying the existing source code a bit.

If you have the time to take a look at doing this I think it would be a great improvement to the project 😄 We could always break this down into a few small PRs. If I create a feature branch for this, we could review code into that feature branch in small chunks as part of a larger project, then merge the branch into master once the full feature is done, WDYT?

If we do that we should probably come up with a plan and break it into small tasks before we get started

**Comment by ploxiln** [comment]
> I do also like the idea of FileSystemCookieStore for those who don't want to set up a redis but alas, you are correct, reaping sessions would be difficult.

I think an in-process cookie store - a `sync.Map` or something like that - may be good enough for the simplest single-instance use case. If the service restarts, people have to re-sign-in, but maybe that's less often than once a week anyway. Just iterating over the whole map looking for expired keys every 60 seconds may be good enough for small scale usage.

**Comment by brianv0** [comment]
I pushed the code around a bit: ae14df4.

It's passing all the tests, but it's not as clean as I'd like.

**Comment by brianv0** [comment]
Another commit:  lsst-dm/oauth2_proxy@5096db0  

You end up with an interface like this:

```
type Store interface {
	Store(requestCookie *http.Cookie, value string, expires time.Time, cookieMaker func(string) *http.Cookie) ([]*http.Cookie, error)
	Load(request *http.Request) (string, error)
	Clear(requestCookie *http.Cookie) (bool, error)
}
```

* `Store` stores a cookie in a backend (if needed) and returns the appropriate cookies to the client. It's able to encapsulate the cookie splitting.
* `Load` takes in the HTTP request and gives the cookie value back, which can be verified and turned into a session.
* `Clear` clears the cookie from the backend. This is a no-op for normal cookies

I think that's a little better than the previous one (both commits are in the experimental_cookie_store branch). I'm not sure they are substantially better than c3e30ca though. It's kind of hard to get good separation of concerns on this because you need some cookie handling/making logic in oauthproxy.go for the CSRF cookie. It's not quite clear what the level of abstraction on the interface should be (request, cookie[s], values/encoded session).

**Comment by JoelSpeed** [comment]
I've been having a think about this over the last couple of days and I'm leaning towards an abstraction at the session level, so the interface would be `SessionStore` and look something like below:

```
type SessionStore interface {
	SaveSession(rw http.ResponseWriter, req *http.Request, s *providers.SessionState) (error)
	LoadSession(req *http.Request) (*providers.SessionState, time.Duration, error)
	ClearSession(rw http.ResponseWriter, req *http.Request) (error)
}
```

At this point it is up to each implementation of the SessionStore to handle setting their own cookies and removes all the extra cookie based logic from the main loop.

To me this seems like the right level because, up until this point, the cookie is being used as a session store. With the Redis store, the cookie it sets is part of its implementation of session storage IMO. I'm not 100% sure how much duplication of logic there will be in logic at this point? Should just be adding the cookie to the upstream? I imagine there will be helper utils for the session stores.

If we take this approach, I think it might be worth moving the `SessionState` type into a `session` package with the stores inside this.

I'd be interested to see if @ploxiln has a say on where to make the abstraction since he knows the codebase better than me. @syscll Do you have any thoughts?

I may have a go at mocking this up over the weekend to see what it looks like 🤔

**Comment by ploxiln** [comment]
The session-oriented interface you propose makes sense to me - so that if one implementation needs to set multiple cookies, that is fully handled behind the interface.

**Comment by brianv0** [comment]
I did some work on this, I will try to work on it some more today and show some changes.

I don't understand what the `time.Duration` is for though, Cookie Validation? I think the cookie validation has to be pushed into the SessionStore with that interface.

It may also be related to #65, since that's effectively loading a mock session based on an incoming JWT. I thought about this over the weekend and I think it's probably advisable some of #65's code move into oidc.go or in a theoretical jwt.go.

A bit of fair warning too - I have been using the `extra-jwt-issuers` and the redis store (+ ticket concept) to actually create issue API tokens. I've found this very effective and lightweight, and much stronger than the HTTP Basic technique, since the issues tokens can have much more information in them.

If you guys would like to do a zoom/skype/slack/etc... some time to discuss this I'd be more than happy.

**Comment by timothy-spencer** [comment]
Hey, dunno if this is the right place, but if you are restructuring session state, it would be cool to find some way to start that in the auth step rather than the code redemption step, because then you can create and use a [nonce](https://openid.net/specs/openid-connect-core-1_0.html#NonceNotes) properly.

That said, maybe there's an alternate way to store this (like a cookie?) that would be better.  I haven't looked into this too hard.  I have a [todo](https://github.com/pusher/oauth2_proxy/blob/master/providers/logingov.go#L23-L24) for this, so if you ended up finding a great way to do this, I'd love to add a nonce to it in the auth step.

**Comment by brianv0** [comment]
On the `time.Duration`, returned by `LoadSession `, it's the `sessionAge`, but it's not actually part of the `SessionState` object. I think I convinced myself that made sense at one point, but I'm not so confident about that anymore.

**Comment by JoelSpeed** [comment]
> If you guys would like to do a zoom/skype/slack/etc... some time to discuss this I'd be more than happy.

I think this would be a great idea. Happy to set up a zoom meeting, I work 10-6pm London time normally though can do an evening call normally, where is everyone else? I'll set up an email thread for this i think 🤔 

I've also spent a little time today and over the weekend working on some ideas of how we should structure this and some tests I'd like to see. The changes I'm thinking of are fairly substantial so I'm still leaning towards setting up a larger feature branch and having the redis store PR'd into that branch or waiting for this branch to be merged, do you think that would be ok @brianv0?

I've pushed my work to the [session-store](https://github.com/pusher/oauth2_proxy/tree/session-store) branch if you'd like to see what I'm working on

**Comment by brianv0** [comment]
I think a feature branch is fine. 

I moved around the code today too, and I think this shouldn't be too much more work, I'll take a look at your branch a bit later.

I'm in California, but I could easily make 4-6PM London time.

**Comment by syscll** [comment]
Also happy to jump on a call if needed. I'm available 10-6pm London time.

**Comment by JoelSpeed** [comment]
I've implemented #147 and #148 towards this project effort. Comments and feedback welcome!

**Comment by JoelSpeed** [comment]
@brianv0 Happy to close this now #155 is merged?
